From 398f377fbce6cf7c994d525ab0471cb5bc8d832c Mon Sep 17 00:00:00 2001
From: Daniel Borkmann <dborkman@redhat.com>
Date: Tue, 22 Jan 2013 16:42:44 +0100
Subject: [PATCH 41/53] sctp_darn: fix __u64 format string issues on ppc64 /
 x86_64
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

On 64-bit powerpc, __u64 is defined to be unsigned long rather than
unsigned long long.  This causes compiler warnings every time we
print a __u64 value with %PRIu64.

Since we cannot change or define our own u64 type in this case,
because it is within a structiure from the kernel, we just cast it
to uint64_t.

This behaviour can be reduced down to the following program:

  #include <stdio.h>
  #include <inttypes.h>
  #include <netinet/sctp.h>

  int main(void)
  {
          __u64 foo = 1;

          printf("%" PRIu64 "\n", foo);
          printf("%" PRIu64 "\n", (uint64_t) foo);

          printf("__u64:%zu versus uint64_t:%zu\n", sizeof(__u64), sizeof(uint64_t));

          return 0;
  }

On x86_64:

  $ gcc foo.c -Wall -O2
  foo.c: In function ‘main’:
  foo.c:9:2: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 2 has type ‘__u64’ [-Wformat]
  $ ./a.out
  [...]
  __u64:8 versus uint64_t:8

On ppc64:

  $ gcc -Wall -O2 foo.c
  $ ./a.out
  [...]
  __u64:8 versus uint64_t:8

So simply cast it in this case to uint64_t.

Signed-off-by: Daniel Borkmann <dborkman@redhat.com>
---
 src/apps/sctp_darn.c | 30 +++++++++++++++---------------
 1 file changed, 15 insertions(+), 15 deletions(-)

diff --git a/src/apps/sctp_darn.c b/src/apps/sctp_darn.c
index 9970cfb..527a4fc 100644
--- a/src/apps/sctp_darn.c
+++ b/src/apps/sctp_darn.c
@@ -2258,21 +2258,21 @@ get_assocstats_func(int sk, sctp_assoc_t assoc_id)
 		return error;
 	}
 
-	printf("Retransmitted Chunks: %" PRIu64 "\n", stats.sas_rtxchunks);
-	printf("Gap Acknowledgements Received: %" PRIu64 "\n", stats.sas_gapcnt);
-	printf("TSN received > next expected: %" PRIu64 "\n", stats.sas_outofseqtsns);
-	printf("SACKs sent: %" PRIu64 "\n", stats.sas_osacks);
-	printf("SACKs received: %" PRIu64 "\n", stats.sas_isacks);
-	printf("Control chunks sent: %" PRIu64 "\n", stats.sas_octrlchunks);
-	printf("Control chunks received: %" PRIu64 "\n", stats.sas_ictrlchunks);
-	printf("Ordered data chunks sent: %" PRIu64 "\n", stats.sas_oodchunks);
-	printf("Ordered data chunks received: %" PRIu64 "\n", stats.sas_iodchunks);
-	printf("Unordered data chunks sent: %" PRIu64 "\n", stats.sas_ouodchunks);
-	printf("Unordered data chunks received: %" PRIu64 "\n", stats.sas_iuodchunks);
-	printf("Dups received (ordered+unordered): %" PRIu64 "\n", stats.sas_idupchunks);
-	printf("Packets sent: %" PRIu64 "\n", stats.sas_opackets);
-	printf("Packets received: %" PRIu64 "\n", stats.sas_ipackets);
-	printf("Maximum Observed RTO this period: %" PRIu64 " - Transport: ", stats.sas_maxrto);
+	printf("Retransmitted Chunks: %" PRIu64 "\n", (uint64_t) stats.sas_rtxchunks);
+	printf("Gap Acknowledgements Received: %" PRIu64 "\n", (uint64_t) stats.sas_gapcnt);
+	printf("TSN received > next expected: %" PRIu64 "\n", (uint64_t) stats.sas_outofseqtsns);
+	printf("SACKs sent: %" PRIu64 "\n", (uint64_t) stats.sas_osacks);
+	printf("SACKs received: %" PRIu64 "\n", (uint64_t) stats.sas_isacks);
+	printf("Control chunks sent: %" PRIu64 "\n", (uint64_t) stats.sas_octrlchunks);
+	printf("Control chunks received: %" PRIu64 "\n", (uint64_t) stats.sas_ictrlchunks);
+	printf("Ordered data chunks sent: %" PRIu64 "\n", (uint64_t) stats.sas_oodchunks);
+	printf("Ordered data chunks received: %" PRIu64 "\n", (uint64_t) stats.sas_iodchunks);
+	printf("Unordered data chunks sent: %" PRIu64 "\n", (uint64_t) stats.sas_ouodchunks);
+	printf("Unordered data chunks received: %" PRIu64 "\n", (uint64_t) stats.sas_iuodchunks);
+	printf("Dups received (ordered+unordered): %" PRIu64 "\n", (uint64_t) stats.sas_idupchunks);
+	printf("Packets sent: %" PRIu64 "\n", (uint64_t) stats.sas_opackets);
+	printf("Packets received: %" PRIu64 "\n", (uint64_t) stats.sas_ipackets);
+	printf("Maximum Observed RTO this period: %" PRIu64 " - Transport: ", (uint64_t) stats.sas_maxrto);
 	print_sockaddr((struct sockaddr *)&stats.sas_obs_rto_ipaddr);
 	printf("\n");
 
-- 
1.7.11.7

